File: cross-platform/src/core/audioAssetCatalog.ts
/**
 * A retained recording used by a session. Applicability is authored in JSON;
 * playback never switches on a Focus key or on the asset's historical wave.
 */
export type AudioAssetRole = "resonantTuning" | "returnSignal";
export type AudioAssetFit = "once" | "cropOrLoop";

export interface AudioAsset {
  id: string;
  role: AudioAssetRole;
  /** Path relative to `library/`, kept relative so an installed library can
   *  move without rewriting its catalog. */
  file: string;
  levels: string[];
  bytes: number;
  sha256: string;
  seconds: number;
  sampleRate: number;
  channels: number;
  fit: AudioAssetFit;
  crossfadeSeconds: number;
  edgeFadeSeconds: number;
  gain: number;
  source: string;
  notes: string;
}

/** `"*"` means every station. Written as a wildcard rather than as an empty
 *  list, because a check requires applicability to be *stated*: an absent
 *  list is an omission, and reading one as "everywhere" would let a
 *  genuinely level-specific asset become universal by having a field left
 *  blank. */
export const everyLevel = "*";

export const assetApplies = (asset: AudioAsset, level: string): boolean =>
  asset.levels.includes(everyLevel) || asset.levels.includes(level);

/** Deliberately not `path.join`: Node's join normalizes `..` segments away,
 *  where Swift's `URL.appending(path:)` does not. This URL is unvalidated —
 *  that is exactly why `assetHasSafeRelativePath` exists as a separate
 *  check — so silently collapsing an unsafe path here would hide the very
 *  thing that check is for. `appending(path:)` does strip one leading `/`
 *  from the appended component rather than re-rooting the URL at it, which
 *  a plain template literal would not do on its own. */
export const assetURL = (asset: AudioAsset, root: string): string =>
  `${root}/library/${asset.file.replace(/^\/+/, "")}`;

/** Catalog paths are data, but they are still confined to the library.
 *  Swift's `split(separator:omittingEmptySubsequences:false)` keeps every
 *  segment including empties; JS's plain `split("/")` already does the same
 *  for a single-character separator. */
export const assetHasSafeRelativePath = (asset: AudioAsset): boolean =>
  asset.file !== "" && !asset.file.startsWith("/") && !asset.file.split("/").includes("..");

/** Required fields throw on absence in Swift (`decode`, not
 *  `decodeIfPresent`), which fails the *whole catalog's* decode — one
 *  malformed entry makes every asset unreadable rather than dropping quietly.
 *  Optional fields fall back the same way every other hand-editable file
 *  here does. */
export function decodeAudioAsset(raw: unknown): AudioAsset {
  const o = (raw ?? {}) as Record<string, unknown>;
  if (typeof o.id !== "string") throw new Error("audio asset missing id");
  const role = o.role === "resonantTuning" || o.role === "returnSignal" ? o.role : undefined;
  if (role === undefined) throw new Error(`audio asset ${o.id} has an unrecognised role`);
  if (typeof o.file !== "string") throw new Error(`audio asset ${o.id} missing file`);
  return {
    id: o.id, role, file: o.file,
    levels: Array.isArray(o.levels) ? o.levels.filter((l): l is string => typeof l === "string") : [],
    bytes: typeof o.bytes === "number" ? o.bytes : 0,
    sha256: typeof o.sha256 === "string" ? o.sha256 : "",
    seconds: typeof o.seconds === "number" ? o.seconds : 0,
    sampleRate: typeof o.sampleRate === "number" ? o.sampleRate : 0,
    channels: typeof o.channels === "number" ? o.channels : 0,
    fit: o.fit === "cropOrLoop" ? "cropOrLoop" : "once",
    crossfadeSeconds: typeof o.crossfadeSeconds === "number" ? o.crossfadeSeconds : 0,
    edgeFadeSeconds: typeof o.edgeFadeSeconds === "number" ? o.edgeFadeSeconds : 1,
    gain: typeof o.gain === "number" ? o.gain : 1,
    source: typeof o.source === "string" ? o.source : "",
    notes: typeof o.notes === "string" ? o.notes : "",
  };
}

export interface AudioAssetCatalog { version: number; distribution: string; assets: AudioAsset[] }

/** A catalog with even one malformed asset fails to decode entirely, the same
 *  as the Swift original: `[AudioAsset]` decodes element by element and
 *  Swift's array `Decodable` conformance does not skip a throwing element. */
export function decodeAudioAssetCatalog(raw: unknown): AudioAssetCatalog {
  const o = (raw ?? {}) as Record<string, unknown>;
  const rawAssets = Array.isArray(o.assets) ? o.assets : [];
  return {
    version: typeof o.version === "number" ? o.version : 1,
    distribution: typeof o.distribution === "string" ? o.distribution : "private",
    assets: rawAssets.map(decodeAudioAsset),
  };
}

export const matchesAsset = (catalog: AudioAssetCatalog, role: AudioAssetRole, level: string): AudioAsset[] =>
  catalog.assets.filter(a => a.role === role && assetApplies(a, level));

File: Sources/GatewayCore/AudioAssetCatalog.swift
import Foundation

public enum AudioAssetRole: String, Codable, CaseIterable, Sendable {
    case resonantTuning
    case returnSignal
}

public enum AudioAssetFit: String, Codable, Sendable {
    case once
    case cropOrLoop
}

/// A retained recording used by a session. Applicability is authored in JSON;
/// playback never switches on a Focus key or on the asset's historical wave.
public struct AudioAsset: Codable, Identifiable, Equatable, Sendable {
    public var id: String
    public var role: AudioAssetRole
    /// Path relative to `library/`, kept relative so an installed library can
    /// move without rewriting its catalog.
    public var file: String
    public var levels: [String]
    public var bytes: Int64
    public var sha256: String
    public var seconds: Double
    public var sampleRate: Double
    public var channels: Int
    public var fit: AudioAssetFit
    public var crossfadeSeconds: Double
    public var edgeFadeSeconds: Double
    public var gain: Double
    public var source: String
    public var notes: String

    public init(id: String, role: AudioAssetRole, file: String, levels: [String],
                bytes: Int64, sha256: String, seconds: Double,
                sampleRate: Double, channels: Int, fit: AudioAssetFit = .once,
                crossfadeSeconds: Double = 0, edgeFadeSeconds: Double = 1,
                gain: Double = 1, source: String = "",
                notes: String = "") {
        self.id = id; self.role = role; self.file = file; self.levels = levels
        self.bytes = bytes; self.sha256 = sha256; self.seconds = seconds
        self.sampleRate = sampleRate; self.channels = channels
        self.fit = fit; self.crossfadeSeconds = crossfadeSeconds
        self.edgeFadeSeconds = edgeFadeSeconds; self.gain = gain
        self.source = source; self.notes = notes
    }

    /// Whether this asset may be used at `level`.
    ///
    /// **`"*"` means every station.** The three resonant
    /// tuning assets carry a real constraint -- Wave I, Wave V and Wave VII
    /// are different recordings for different levels -- but the wake-up
    /// warble is one recording that is not level-specific at all. Its list
    /// enumerated the seventeen levels `levels.json` happened to document,
    /// which was harmless only while a session could not end anywhere else.
    ///
    /// Continuous mode ended that: a journey can arrive at any of the
    /// forty-two stations on the ladder. Arriving at Focus 13 with the list
    /// as an allowlist left the listener held at the station with **no way to
    /// be counted back** -- "Return to waking" refused, because the wake-up
    /// signal it needs did not admit to applying there.
    ///
    /// Written as a wildcard rather than as an empty list, because gfcheck
    /// requires applicability to be *stated*: an absent list is an omission,
    /// and reading one as "everywhere" would let a genuinely level-specific
    /// asset become universal by having a field left blank.
    public static let everyLevel = "*"

    public func applies(to level: String) -> Bool {
        levels.contains(Self.everyLevel) || levels.contains(level)
    }

    public func url(in root: URL) -> URL {
        root.appending(path: "library").appending(path: file)
    }

    /// Catalog paths are data, but they are still confined to the library.
    public var hasSafeRelativePath: Bool {
        !file.isEmpty && !file.hasPrefix("/")
            && !file.split(separator: "/", omittingEmptySubsequences: false)
                .contains("..")
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        role = try c.decode(AudioAssetRole.self, forKey: .role)
        file = try c.decode(String.self, forKey: .file)
        levels = try c.decodeIfPresent([String].self, forKey: .levels) ?? []
        bytes = try c.decodeIfPresent(Int64.self, forKey: .bytes) ?? 0
        sha256 = try c.decodeIfPresent(String.self, forKey: .sha256) ?? ""
        seconds = try c.decodeIfPresent(Double.self, forKey: .seconds) ?? 0
        sampleRate = try c.decodeIfPresent(Double.self, forKey: .sampleRate) ?? 0
        channels = try c.decodeIfPresent(Int.self, forKey: .channels) ?? 0
        fit = try c.decodeIfPresent(AudioAssetFit.self, forKey: .fit) ?? .once
        crossfadeSeconds = try c.decodeIfPresent(Double.self, forKey: .crossfadeSeconds) ?? 0
        edgeFadeSeconds = try c.decodeIfPresent(Double.self, forKey: .edgeFadeSeconds) ?? 1
        gain = try c.decodeIfPresent(Double.self, forKey: .gain) ?? 1
        source = try c.decodeIfPresent(String.self, forKey: .source) ?? ""
        notes = try c.decodeIfPresent(String.self, forKey: .notes) ?? ""
    }
}

public struct AudioAssetCatalog: Codable, Equatable, Sendable {
    public var version: Int
    public var distribution: String
    public var assets: [AudioAsset]

    public init(version: Int = 1, distribution: String = "private",
                assets: [AudioAsset] = []) {
        self.version = version; self.distribution = distribution; self.assets = assets
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        version = try c.decodeIfPresent(Int.self, forKey: .version) ?? 1
        distribution = try c.decodeIfPresent(String.self, forKey: .distribution) ?? "private"
        assets = try c.decodeIfPresent([AudioAsset].self, forKey: .assets) ?? []
    }

    public static func load(root: URL) throws -> AudioAssetCatalog {
        let url = root.appending(path: "library/audio-assets.json")
        return try JSONDecoder().decode(Self.self, from: Data(contentsOf: url))
    }

    public func matches(role: AudioAssetRole, level: String) -> [AudioAsset] {
        assets.filter { $0.role == role && $0.applies(to: level) }
    }
}
